### PR TITLE
futilbackend extended to support fixedpoint

### PR DESCRIPTION
**Pull request by yoonachang**
supports  fixedpoint constants and arithmetic // It does not support addition of different bitwidths yet

### CODE REVIEW

File: src/main/scala/backends/futil/FutilAst.scala [hunk 1]
```diff
@@ -317,11 +318,30 @@ object Stdlib {
...
+    Futil.CompInst("fixed_p_std_const", List(width, int_bit, frac_bit, value1, value2))
+
+  def fxd_p_op(op: String,
+      width: Int,
+      int_bit: Int,
+      frac_bit: Int
+  ): Futil.CompInst =
+    Futil.CompInst(s"fixed_p_std_$op", List(width, int_bit, frac_bit))
 
+/////
```
Comment (rachitnigam) [inline src/main/scala/backends/futil/FutilAst.scala:339]: Unnecessary change

File: src/main/scala/backends/futil/FutilAst.scala [hunk 2]
```diff
@@ -317,11 +318,30 @@ object Stdlib {
...
   val staticTimingMap: Map[String, Option[Int]] = Map(
     "sqrt" -> Some(17),
     "mult" -> Some(3),
     "div" -> None,
     "mod" -> None
   )
+
```
Comment (rachitnigam) [inline src/main/scala/backends/futil/FutilAst.scala:346]: Unnecessary change

File: src/main/scala/backends/futil/FutilBackend.scala [hunk 3]
```diff
@@ -155,19 +166,57 @@ private class FutilBackendHelper {
       val reg = LibDecl(CompVar(s"${d.id}"), Stdlib.register(size))
       List(reg)
     }
+    case TFixed(ltotal, _ , unsigned)=>{
+      assert(unsigned, NotImplemented("Generating signed fixed points", d.pos))
+      val reg = LibDecl(CompVar(s"${d.id}"), Stdlib.register(ltotal))
+      List(reg)
+    } // my change
```
Comment (rachitnigam) [inline src/main/scala/backends/futil/FutilBackend.scala:173]: stray comment

File: src/main/scala/backends/futil/FutilBackend.scala [hunk 4]
```diff
@@ -79,6 +89,7 @@ private class FutilBackendHelper {
         assert(unsigned, NotImplemented("Arrays of signed integers."))
         size
       }
+      //later expand fixd point
```
Comment (rachitnigam) [inline src/main/scala/backends/futil/FutilBackend.scala:92]: If you add a TODO comment, do this:
```
TODO(yoonachang): Later expand fixed point. <Some information on how you think this should be done>.
```

File: src/main/scala/backends/futil/FutilBackend.scala [hunk 5]
```diff
@@ -39,12 +39,22 @@ private class FutilBackendHelper {
...
-  def bitsForType(t: Option[Type], pos: Position): Int = {
+  // Q - extend fixedpints ?
+  def bitsForType(t: Option[Type], pos: Position): (Int, Option[Int]) = {
     t match {
-      case Some(TSizedInt(width, _)) => width
-      case Some(_: TBool) => 1
-      case Some(_: TVoid) => 0
+      case Some(TSizedInt(width, _)) => (width, None)
+      case Some(TFixed(t, i, _)) => (t, Some(i)) // my change
+      // case Some(_: TRational) =>
```
Comment (rachitnigam) [inline src/main/scala/backends/futil/FutilBackend.scala:55]: Stray comment

File: src/main/scala/backends/futil/FutilBackend.scala [hunk 6]
```diff
@@ -155,19 +166,57 @@ private class FutilBackendHelper {
...
+    EmitOutput(
+      comp.id.port("out"),
+      ConstantPort(1, 1),
+      struct ++ e1Out.structure ++ e2Out.structure,
+      for (d1 <- e1Out.delay; d2 <- e2Out.delay)
+        yield d1 + d2
+    )
+  }
+
   def emitBinop(compName: String, e1: Expr, e2: Expr)(
```
Comment (rachitnigam) [inline src/main/scala/backends/futil/FutilBackend.scala:213]: Missing description.

File: src/main/scala/backends/futil/FutilBackend.scala [hunk 7]
```diff
@@ -253,8 +304,24 @@ private class FutilBackendHelper {
...
+        // if the typ of e1 is  fixed points do fixed point binary operation 
+        if (isfixedpoint(e1.typ)) {
+          val compName =op.op match {
+            case "+" => "add"
+            case "-" => "sub"
+            case "*" => "mult"
+            case "/" => "div"
+            case x =>
+              throw NotImplemented(
+                s"Futil backend does not support '$x' yet.",
```
Comment (rachitnigam) [inline src/main/scala/backends/futil/FutilBackend.scala:316]: In the error message should say that this is the fixed-point multiply.

File: file-tests/should-run/simpletest.fuse [hunk 8]
```diff
@@ -0,0 +1,12 @@
+//let t: bit<16> =2;
```
Comment (rachitnigam) [inline file-tests/should-run/simpletest.fuse:1]: Can you remove this test since it exists in the FuTIL repository?

Comment (yoonachang) [reply file-tests/should-run/simpletest.fuse:1]: Yeah so I also have to remove the test for matadd-fixed-points.fuse? Or just duplicate that for futil repository. Also I'll just push the updated version that supports arrays.

Comment (yoonachang) [reply file-tests/should-run/simpletest.fuse]: I removed the simpletest and dot product/ fixed point array part is complete
